ksql-engine/src/main/java/io/confluent/ksql/KsqlEngine.java

```diff
@@ -761,28 +678,17 @@ private static Query addInto(
   private static final class SandboxedExecutionContext implements KsqlExecutionContext {
 
     private final EngineContext engineContext;
-    private final Map<QueryId, PersistentQueryMetadata> persistentQueries;
 
-    private static SandboxedExecutionContext create(
-        final EngineContext engineContext,
-        final Map<QueryId, PersistentQueryMetadata> persistentQueries
-    ) {
-      final EngineContext sandboxed = EngineContext.create(
-          SandboxedServiceContext.create(engineContext.serviceContext),
-          engineContext.metaStore.copy(),
-          query -> {
-          } // Do nothing on query close.
+    private SandboxedExecutionContext(final EngineContext sourceContext) {
+      this.engineContext = EngineContext.create(
+          SandboxedServiceContext.create(sourceContext.serviceContext),
+          sourceContext.metaStore.copy(),
+          sourceContext.queryIdGenerator.peek(),
+          this::unregisterQuery
       );
 
-      return new SandboxedExecutionContext(sandboxed, persistentQueries);
-    }
-
-    private SandboxedExecutionContext(
-        final EngineContext engineContext,
-        final Map<QueryId, PersistentQueryMetadata> persistentQueries
-    ) {
-      this.engineContext = Objects.requireNonNull(engineContext, "engineContext");
-      this.persistentQueries = ImmutableMap.copyOf(persistentQueries);
+      sourceContext.persistentQueries.forEach((queryId, query) ->
+          engineContext.registerQuery(query.copyWith(this::unregisterQuery)));
```

Comment:
I'm still strongly in the other camp ;)

After building a query, you should just be able to close it, and for that close to do the right thing.

Reply:
One day this will come back and bite us! 😉

Reply:
I'm really struggling to understand your reasons for thinking this is a bad pattern.  This is a pretty standard pattern.  

Consider an example of a FileSystem that allows you to get a InputStream on file content and which tracks which files are currently open:

```java
InputStream s = fileSytem.openInputStream(someFile);
... // do stuff with s.
s.close();
```

In this example, you'd expect the stream `close()` method to ensure the tracking info in the fileSystem is updated.  Requiring something:

```java
InputStream s = fileSytem.openInputStream(someFile);
... // do stuff with s.
fileSytem.closeStream(s);
```

Would just be weird.  Also, this second pattern requires that you need to pass around not only the created stream, but also the fileSystem object so that you can later close the stream.

In the case of the engine / query code... the engine creates the query.  The query is belongs to the engine.  As such, closing the query _should_ update the state of the engine.  Also consider, (somewhat convoluted for engine / query example)

```java
Query query = engineOne.execute(stmt);
engineTwo.closeQuery(query);
```

By requiring the lifecycle of a created resource to be controlled by the creating entity, you actually make the API easier to miss use.

Reply:
I think I can be convinced of this one (and I'm not just saying that because I want to hedge my bets on the side-effects discussion 😛). Perhaps my concern comes from some mingling of state that should be separate that I can't easily identify. With the case of filesystems, I'm pretty darn sure that if I call `s1.close()` on a stream `s1` that was opened with filesystem `f1`, filesystem `f2` won't be randomly affected. 

The way we had ksqlEngine, however, something like similar is possible. After your unification PRs this has dramatically improved, but in previous code paths if I were to create a query using a sandboxed service context, it was possible that it was later closed using the real service context because it was implicitly referencing instance variables of ksqlEngine.

I guess you're right that this pattern of close callbacks is not bad, but something else in there is causing my code senses to tingle, I just haven't figured it out yet.

Reply:
Spidey senses tingling eh?

> I'm pretty darn sure that if I call `s1.close()` on a stream `s1` that was opened with filesystem `f1`, filesystem `f2` won't be randomly affected.

Yep, and the same should be true for the engine.  Sure, there could be a bug. Just as their could be a bug in the filesystem example that meant your assumption didn't stand there.  But that doesn't invalidate the pattern.

Anyway... moving on ;)